capstone/src/test/java/com/google/sps/servlets/MarkCheckboxTest.java

```diff
@@ -0,0 +1,135 @@
+import com.google.sps.Objects.Challenge;
+
+public class MarkCheckboxTest {
+  private static final String USER_EMAIL = "test@test.com";
+  private static final String USER_ID = "test";
+  private static final String CHECKBOX_ID = "1";
+  private static final String CHECKED = "true";
+  private static final String TYPE = "Challenge";
+  private static final String NEW_CHALLENGE = "Run";
+
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(
+              new LocalDatastoreServiceTestConfig()
+                  .setDefaultHighRepJobPolicyUnappliedJobPercentage(0),
+              new LocalUserServiceTestConfig())
+          .setEnvEmail(USER_EMAIL)
+          .setEnvIsLoggedIn(true)
+          .setEnvAuthDomain("gmail.com")
+          .setEnvAttributes(
+              new HashMap(
+                  ImmutableMap.of(
+                      "com.google.appengine.api.users.UserService.user_id_key", USER_ID)));
+
+  @Mock private HttpServletRequest mockRequest;
+  @Mock private HttpServletResponse mockResponse;
+  private StringWriter responseWriter;
+  private DatastoreService datastore;
+  private MarkCheckboxServlet markCheckboxServlet;
+  private ChallengeServlet challengeServlet;
+
+  @Before
+  public void setUp() throws IOException {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Set up a fake HTTP response.
+    responseWriter = new StringWriter();
+    when(mockResponse.getWriter()).thenReturn(new PrintWriter(responseWriter));
+
+    markCheckboxServlet = new MarkCheckboxServlet();
+    challengeServlet = new ChallengeServlet();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+    responseWriter = null;
+    datastore = null;
+    markCheckboxServlet = null;
+    challengeServlet = null;
+  }
+
+  @Test
+  public void doPost_userLoggedIn_validChallenge() throws IOException, EntityNotFoundException {
+    when(mockRequest.getParameter("name")).thenReturn(NEW_CHALLENGE);
+
+    challengeServlet.doPost(mockRequest, mockResponse);
+
+    when(mockRequest.getParameter("id")).thenReturn(CHECKBOX_ID);
+    when(mockRequest.getParameter("checked")).thenReturn(CHECKED);
+    when(mockRequest.getParameter("type")).thenReturn(TYPE);
+
+    markCheckboxServlet.doPost(mockRequest, mockResponse);
+
+    Key key = KeyFactory.createKey(TYPE, Long.parseLong(CHECKBOX_ID));
+    Entity entity = datastore.get(key);
+    Challenge challenge = Challenge.fromEntity(entity);
+
+    assert challenge.getUsersCompleted().size() == 1;
+    assertEquals(challenge.getUsersCompleted().get(0), USER_ID);
+  }
+
+  @Test(expected = EntityNotFoundException.class)
+  public void doPost_userNotLoggedIn() throws IOException, EntityNotFoundException {
+    helper.setEnvIsLoggedIn(false);
+    when(mockRequest.getParameter("name")).thenReturn(NEW_CHALLENGE);
+
+    challengeServlet.doPost(mockRequest, mockResponse);
+
```

Comment:
Tests are usually preferred to be in three blocks:

// Arrange

// Act

// Assert

Right now, it's hard for me to understand why you have to `challengeServlet.doPost` before doing your `markCheckboxServlet.doPost`. I recommend one of two suggestions:

1. Seeing multiple blocks here, as well as a couple of repeated lines of code in multiple tests suggests that adding a helper function for testing may make this read more clearly:

```java
helper.setEnvIsLoggedIn(false);
setupChallenge();
when(mockRequest.getParameter("id")....
```

2. If you have to call another file's doPost in order to make your unit test work, it may not be a unit test. Consider mocking some behavior that MarkCheckbox is expecting from the database.